opentelemetry-sdk/src/opentelemetry/sdk/metrics/__init__.py

```diff
@@ -356,12 +356,48 @@ def __init__(
     ):
         self.instrumentation_info = instrumentation_info
         self.processor = Processor(source.stateful, source.resource)
-        self.metrics = set()
-        self.observers = set()
+        self.metrics = dict()
+        self.observers = dict()
         self.metrics_lock = threading.Lock()
         self.observers_lock = threading.Lock()
         self.view_manager = ViewManager()
 
+    def _register_metric(
+        self,
+        instrument: Union[
+            metrics_api.Counter,
+            metrics_api.UpDownCounter,
+            metrics_api.ValueRecorder,
+        ],
```

Comment:
I think `metrics_api.Metric` would work here and then `metrics_api.Observer` below